field/src/extension/binomial_extension.rs

```diff
@@ -0,0 +1,364 @@
+
+    const ZERO: Self = Self([F::ZERO; D]);
+    const ONE: Self = Self(field_to_array::<F, D>(F::ONE));
+    const TWO: Self = Self(field_to_array::<F, D>(F::TWO));
+    const NEG_ONE: Self = Self(field_to_array::<F, D>(F::NEG_ONE));
+
+    fn from_bool(b: bool) -> Self {
+        F::from_bool(b).into()
+    }
+
+    fn from_canonical_u8(n: u8) -> Self {
+        F::from_canonical_u8(n).into()
+    }
+
+    fn from_canonical_u16(n: u16) -> Self {
+        F::from_canonical_u16(n).into()
+    }
+
+    fn from_canonical_u32(n: u32) -> Self {
+        F::from_canonical_u32(n).into()
+    }
+
+    /// Convert from `u64`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_u64(n: u64) -> Self {
+        F::from_canonical_u64(n).into()
+    }
+
+    /// Convert from `usize`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_usize(n: usize) -> Self {
+        F::from_canonical_usize(n).into()
+    }
+
+    fn from_wrapped_u32(n: u32) -> Self {
+        F::from_wrapped_u32(n).into()
+    }
+
+    fn from_wrapped_u64(n: u64) -> Self {
+        F::from_wrapped_u64(n).into()
+    }
+
+    fn generator() -> Self {
+        Self(F::ext_multiplicative_group_generator())
+    }
+
+    #[inline(always)]
+    fn square(&self) -> Self {
+        // Specialising mul reduces the computation of c1 from 2 muls
+        // and one add to one mul and a shift
+        *self * *self
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Field for BinomialExtensionField<F, D> {
+    type Packing = Self;
+    // Algorithm 11.3.4 in Handbook of Elliptic and Hyperelliptic Curve Cryptography.
+    fn try_inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            return None;
+        }
+
+        // Writing 'a' for self, we need to compute a^(r-1):
+        // r = n^D-1/n-1 = n^(D-1)+n^(D-2)+...+n
+        let mut f = Self::ONE;
+        for _ in 1..D {
+            f = (f * *self).frobenius();
+        }
+
+        // g = a^r is in the base field, so only compute that
+        // coefficient rather than the full product.
+
+        let Self(a) = *self;
+        let Self(b) = f;
+        let mut g = F::ZERO;
+        for i in 1..D {
+            g += a[i] * b[D - i];
+        }
+        g *= F::W;
+        g += a[0] * b[0];
+        debug_assert_eq!(Self::from(g), *self * f);
+
+        Some(f * g.inverse())
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Display for BinomialExtensionField<F, D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let mut result = String::new();
+        for (i, &x) in self.0.iter().enumerate() {
+            if i == 0 {
+                result.push_str(&(format!("{}", x)));
+            } else if i == 1 {
+                result.push_str(&format!("+{}x", x));
+            } else {
+                result.push_str(&format!("+{}x^{}", x, i));
+            }
+        }
+        write!(f, "{}", result)
```

Comment:
I think it would be cleaner to do
```suggestion
        let linear_part = format!("{} + {}*X", self.0[0], self.0[1]); // ok, since D >= 2
        let nonlin_part: String = self.0[2..].iter().zip(2..).map(|(x, i)| format!(" + {x}*X^{i}")).collect();
        write!(f, "{}", linear_part + &nonlin_part)
```

Reply:
updated,thx